File: src/Sems.Api/Modules/Subscriptions/Interfaces/SubscriptionController.cs
using Microsoft.AspNetCore.Mvc;
using Sems.Api.Modules.Subscriptions.Application;
using static Sems.Api.Modules.Subscriptions.Interfaces.SubscriptionResources;

namespace Sems.Api.Modules.Subscriptions.Interfaces;

/// <summary>
/// API REST del bounded context de suscripciones.
///
/// <para>Rutas identicas a las del microservicio en Go. Ojo con el orden de los
/// segmentos: <c>/subscriptions/users/{userId}</c> debe declararse antes que
/// <c>/subscriptions/{subscriptionId}</c> para que "users" no se interprete
/// como un identificador. La restriccion <c>:guid</c> lo garantiza ademas por
/// tipo.</para>
/// </summary>
[ApiController]
[Route("api/v1")]
[Tags("Subscriptions")]
public sealed class SubscriptionController : ControllerBase
{
    private readonly SubscriptionService _service;

    public SubscriptionController(SubscriptionService service) => _service = service;

    // ------------------------------------------------------------------ planes

    /// <summary>Lista los planes disponibles.</summary>
    [HttpGet("subscription-plans")]
    public async Task<List<PlanResource>> Plans() =>
        (await _service.ActivePlansAsync()).Select(PlanResource.From).ToList();

    /// <summary>Obtiene un plan por su identificador.</summary>
    [HttpGet("subscription-plans/{planId:guid}")]
    public async Task<PlanResource> PlanById(Guid planId) =>
        PlanResource.From(await _service.PlanByIdAsync(planId));

    // ----------------------------------------------------------- suscripciones

    /// <summary>Suscripciones de un usuario.</summary>
    [HttpGet("subscriptions/users/{userId}")]
    public async Task<List<SubscriptionResource>> ByUser(string userId) =>
        (await _service.SubscriptionsByUserAsync(userId)).Select(SubscriptionResource.From).ToList();

    /// <summary>Obtiene una suscripcion por su identificador.</summary>
    [HttpGet("subscriptions/{subscriptionId:guid}")]
    public async Task<SubscriptionResource> ById(Guid subscriptionId) =>
        SubscriptionResource.From(await _service.SubscriptionByIdAsync(subscriptionId));

    /// <summary>Crea una suscripcion.</summary>
    [HttpPost("subscriptions")]
    public async Task<ActionResult<SubscriptionResource>> Create(
        [FromBody] CreateSubscriptionRequest request)
    {
        var subscription = await _service.CreateAsync(request.UserId, Guid.Parse(request.PlanId), null);
        return StatusCode(StatusCodes.Status201Created, SubscriptionResource.From(subscription));
    }

    /// <summary>Cancela una suscripcion.</summary>
    [HttpPatch("subscriptions/{subscriptionId:guid}/cancel")]
    public async Task<SubscriptionResource> Cancel(Guid subscriptionId) =>
        SubscriptionResource.From(await _service.CancelAsync(subscriptionId));

    /// <summary>Cambia el plan de una suscripcion.</summary>
    [HttpPatch("subscriptions/{subscriptionId:guid}/change-plan")]
    public async Task<SubscriptionResource> ChangePlan(Guid subscriptionId,
        [FromBody] ChangePlanRequest request) =>
        SubscriptionResource.From(
            await _service.ChangePlanAsync(subscriptionId, Guid.Parse(request.NewPlanId)));
}

File: src/main/java/com/sems/subscriptions/interfaces/rest/SubscriptionController.java
package com.sems.subscriptions.interfaces.rest;

import com.sems.subscriptions.application.SubscriptionService;
import com.sems.subscriptions.interfaces.rest.resources.SubscriptionResources.*;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import java.util.List;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

/**
 * API REST del bounded context de suscripciones.
 *
 * <p>Rutas identicas a las del microservicio en Go. Ojo con el orden de los
 * segmentos: {@code /subscriptions/users/{userId}} debe declararse antes que
 * {@code /subscriptions/{subscriptionId}} para que "users" no se interprete
 * como un identificador.
 */
@Tag(name = "Subscriptions", description = "Planes y suscripciones de usuario")
@RestController
@RequestMapping("/api/v1")
@RequiredArgsConstructor
public class SubscriptionController {

    private final SubscriptionService service;

    // ---------------------------------------------------------------- planes

    @Operation(summary = "Lista los planes disponibles")
    @GetMapping("/subscription-plans")
    public List<PlanResource> plans() {
        return service.activePlans().stream().map(PlanResource::from).toList();
    }

    @Operation(summary = "Obtiene un plan por su identificador")
    @GetMapping("/subscription-plans/{planId}")
    public PlanResource planById(@PathVariable UUID planId) {
        return PlanResource.from(service.planById(planId));
    }

    // --------------------------------------------------------- suscripciones

    @Operation(summary = "Suscripciones de un usuario")
    @GetMapping("/subscriptions/users/{userId}")
    public List<SubscriptionResource> byUser(@PathVariable String userId) {
        return service.subscriptionsByUser(userId).stream().map(SubscriptionResource::from).toList();
    }

    @Operation(summary = "Obtiene una suscripcion por su identificador")
    @GetMapping("/subscriptions/{subscriptionId}")
    public SubscriptionResource byId(@PathVariable UUID subscriptionId) {
        return SubscriptionResource.from(service.subscriptionById(subscriptionId));
    }

    @Operation(summary = "Crea una suscripcion")
    @PostMapping("/subscriptions")
    public ResponseEntity<SubscriptionResource> create(
            @Valid @RequestBody CreateSubscriptionRequest request) {
        var subscription = service.create(request.userId(), UUID.fromString(request.planId()), null);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(SubscriptionResource.from(subscription));
    }

    @Operation(summary = "Cancela una suscripcion")
    @PatchMapping("/subscriptions/{subscriptionId}/cancel")
    public SubscriptionResource cancel(@PathVariable UUID subscriptionId) {
        return SubscriptionResource.from(service.cancel(subscriptionId));
    }

    @Operation(summary = "Cambia el plan de una suscripcion")
    @PatchMapping("/subscriptions/{subscriptionId}/change-plan")
    public SubscriptionResource changePlan(@PathVariable UUID subscriptionId,
                                           @Valid @RequestBody ChangePlanRequest request) {
        return SubscriptionResource.from(
                service.changePlan(subscriptionId, UUID.fromString(request.newPlanId())));
    }
}
